### PR TITLE
Update getjdkbinary to correctly handle Temurin debugimage and static-libs folders

### PR DESCRIPTION
**Pull request by andrew-m-leonard**
Fixes https://github.com/adoptium/aqa-tests/issues/5749

To resolve the above issue, this PR:
- Updates the debugimage "overlay" logic to handle the differing Temurin and IBM/semeru debugimage archive structure, with differing number of top-level folders that need strippping
- Add support for Temurin static-libs uncompressing
- Add safety check to ensure the archive contents are moved correctly and not overlaying within an existing expansion of another archive. If a new type of archive is added this check will fail to ensure we don't corrupt the j2sdk-image content with unknown material

**Grinders:**
- sanity.openjdk from "nightly" : https://ci.adoptium.net/job/Grinder/11331/ - UNSTABLE
  - 1 failure: java/lang/reflect/records/RecordReflectionTest.java
  - This fails irrespective to this PR
- sanity.openjdk from "nightly" on Windows jdk8u : https://ci.adoptium.net/job/Grinder/11336/ - SUCCESS
- Rebuild_Same_JDK_Reproducibility_Test_0 from "upstream" artifacts : https://ci.adoptium.net/job/Grinder/11339/ - SUCCESS

### CODE REVIEW

**Comment by llxia** [comment]
There are other cases that we need to test (i.e., openj9, vmfarm, zOS, etc)

**Comment by llxia** [comment]
Failed on [Windows](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44947/console):

```
00:01:45.147  -rw-r--r--+ 1 jenkins None  68818399 Nov 14 15:41 ibm-semeru-certified-debugimage_x64_windows_JDK11U_2024-11-13-23-01.zip
00:01:45.147  -rw-r--r--+ 1 jenkins None 231312038 Nov 14 15:41 ibm-semeru-certified-jdk_x64_windows_JDK11U_2024-11-13-23-01.zip
00:01:45.147  -rw-r--r--+ 1 jenkins None   7886951 Nov 14 15:41 ibm-semeru-certified-testimage_x64_windows_JDK11U_2024-11-13-23-01.zip
00:01:45.148  drwxrwxr-x+ 1 jenkins None         0 Nov 13 17:06 j2sdk-image
00:01:45.148  drwxr-xr-x+ 1 jenkins None         0 Nov 14 15:41 tmp
00:01:45.148  List files in jdkbinary/tmp folder...
00:01:45.148  total 4
00:01:45.148  drwxrwxr-x+ 1 jenkins None 0 Nov 13 17:03 jdk-11.0.25+9-test-image
00:01:45.148  Moving directory jdk-11.0.25+9-test-image/ to ../openjdk-test-image
00:01:46.209  Uncompressing ibm-semeru-certified-debugimage_x64_windows_JDK11U_2024-11-13-23-01.zip over ./j2sdk-image...
00:01:46.210  unzip:  cannot find or open ibm-semeru-certified-debugimage_x64_windows_JDK11U_2024-11-13-23-01.zip, ibm-semeru-certified-debugimage_x64_windows_JDK11U_2024-11-13-23-01.zip.zip or ibm-semeru-certified-debugimage_x64_windows_JDK11U_2024-11-13-23-01.zip.ZIP.
```

**Comment by smlambert** [comment]
Hi Lan, can you share a public Grinder link (or a rerun in Grinder link from the hyc Jenkins run) so Andrew can test modifications he makes?

**Comment by llxia** [comment]
Openj9 Windows is fine https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/3982/. This problem happened when using the windows zip from the internal artifactory. I do not think rerun in Grinder link from the hyc Jenkins would help in this case.

**Comment by llxia** [comment]
- [vmfarm xlinux](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44954/) - unzip passed. 
- [vmfarm windows](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44955/) - unzip passed. 
- [zos](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44950/) - passed 2/3
   - 1 failed due to 
   
```
00:03:00.861  Moving directory test-images/ to ../openjdk-test-image
00:03:00.861  Uncompressing debug-image.tar.gz over ./j2sdk-image...
00:03:06.899  
00:03:06.900  gzip: write_error():stdout: EDC5140I Broken pipe.
```
   - [rerun](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder_testList_0/2344/)

**Comment by andrew-m-leonard** [comment]
> Failed on [Windows](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44947/console):
> 
> ```
> 00:01:18.078  Uncompressing file: ibm-semeru-certified-jdk_x64_windows_JDK11U_2024-11-13-23-01.zip ...
> 00:01:44.169  List files in jdkbinary folder...
> 00:01:44.170  total 300804
> 00:01:44.170  -rw-r--r--+ 1 jenkins None  68818399 Nov 14 15:41 ibm-semeru-certified-debugimage_x64_windows_JDK11U_2024-11-13-23-01.zip
> 00:01:44.170  -rw-r--r--+ 1 jenkins None 231312038 Nov 14 15:41 ibm-semeru-certified-jdk_x64_windows_JDK11U_2024-11-13-23-01.zip
> 00:01:44.170  -rw-r--r--+ 1 jenkins None   7886951 Nov 14 15:41 ibm-semeru-certified-testimage_x64_windows_JDK11U_2024-11-13-23-01.zip
> 00:01:44.170  drwxr-xr-x+ 1 jenkins None         0 Nov 14 15:41 tmp
> 00:01:44.170  List files in jdkbinary/tmp folder...
> 00:01:44.170  total 4
> 00:01:44.170  drwxrwxr-x+ 1 jenkins None 0 Nov 13 17:06 jdk-11.0.25+9
> 00:01:44.170  Moving directory jdk-11.0.25+9/ to ../j2sdk-image
> 00:01:44.170  Uncompressing file: ibm-semeru-certified-testimage_x64_windows_JDK11U_2024-11-13-23-01.zip ...
> 00:01:45.147  List files in jdkbinary folder...
> 00:01:45.147  total 300808
> 00:01:45.147  -rw-r--r--+ 1 jenkins None  68818399 Nov 14 15:41 ibm-semeru-certified-debugimage_x64_windows_JDK11U_2024-11-13-23-01.zip
> 00:01:45.147  -rw-r--r--+ 1 jenkins None 231312038 Nov 14 15:41 ibm-semeru-certified-jdk_x64_windows_JDK11U_2024-11-13-23-01.zip
> 00:01:45.147  -rw-r--r--+ 1 jenkins None   7886951 Nov 14 15:41 ibm-semeru-certified-testimage_x64_windows_JDK11U_2024-11-13-23-01.zip
> 00:01:45.148  drwxrwxr-x+ 1 jenkins None         0 Nov 13 17:06 j2sdk-image
> 00:01:45.148  drwxr-xr-x+ 1 jenkins None         0 Nov 14 15:41 tmp
> 00:01:45.148  List files in jdkbinary/tmp folder...
> 00:01:45.148  total 4
> 00:01:45.148  drwxrwxr-x+ 1 jenkins None 0 Nov 13 17:03 jdk-11.0.25+9-test-image
> 00:01:45.148  Moving directory jdk-11.0.25+9-test-image/ to ../openjdk-test-image
> 00:01:46.209  Uncompressing ibm-semeru-certified-debugimage_x64_windows_JDK11U_2024-11-13-23-01.zip over ./j2sdk-image...
> 00:01:46.210  unzip:  cannot find or open ibm-semeru-certified-debugimage_x64_windows_JDK11U_2024-11-13-23-01.zip, ibm-semeru-certified-debugimage_x64_windows_JDK11U_2024-11-13-23-01.zip.zip or ibm-semeru-certified-debugimage_x64_windows_JDK11U_2024-11-13-23-01.zip.ZIP.
> ```
> 
> [Unzip](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44948/console) is fine without using this PR

@llxia thank you Lan, I see the issue

**Comment by andrew-m-leonard** [comment]
Re-test Windows using upstream artifacts: https://ci.adoptium.net/job/Grinder/11389/console

**Comment by andrew-m-leonard** [comment]
Re-test s390x upstream artifacts: https://ci.adoptium.net/job/Grinder/11390/

**Comment by andrew-m-leonard** [comment]
> * [vmfarm xlinux](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44954/) - unzip passed.
> 
>     * [vmfarm windows](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44955/) - wip
> 
>     * [zos](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44950/) - passed 2/3
>       
>       * 1 failed due to
> 
> 
> ```
> 00:03:00.861  Moving directory test-images/ to ../openjdk-test-image
> 00:03:00.861  Uncompressing debug-image.tar.gz over ./j2sdk-image...
> 00:03:06.899  
> 00:03:06.900  gzip: write_error():stdout: EDC5140I Broken pipe.
> ```
> 
>     * [rerun](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder_testList_0/2344/) on the same machine fyrec21g - failed with the same error
> 
>     * [rerun without using this PR](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder_testList_0/2345/) on the same machine fyrec21g - no `EDC5140I Broken pipe` error

Hi Lan,
I've fixed the zip problem, but the zOS gzip "Broken pipe" I  can't see any obvious issue,
please can you retry on zOS with and without the PR ?
thanks

**Comment by llxia** [comment]
- [windows](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44961/)
- [vmfarm xlinux](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44962/)
- [vmfarm windows](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44963/)
- [openj9 windows](https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/3989/)
- [zos fyrec21g](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder_testList_0/2346/)

File: get.sh
Comment (sxa) [inline get.sh:452]: I'm curious as to how we're ending up the situation where this function is required, as it seems to add to the compexity somewhat. If we really want to catch the condition where it already exists can we just use somethign like:
```suggestion
							[ ! -r ../openjdk-test-image ] && mv "$jar_dir_name" ../openjdk-test-image || exit 1
```
(If you really need a message that should be easy to add)

File: get.sh
Comment (sxa) [inline get.sh:164]: We discussed this problem with z/OS earlier, but what was the problem that this function solved?

File: get.sh
Comment (andrew-m-leonard) [reply get.sh:452]: catching the condition is key, as I don't think when it was originally added as to the side effect of a new type of archive being added, and trashing a different archives content

File: get.sh
Comment (andrew-m-leonard) [reply get.sh:164]: it removes a folder if the archive has just a single top level folder

Comment (andrew-m-leonard) [reply get.sh:164]: should I call it "stripFolderIfArchiveContainsOnlyOne" ?

File: get.sh
Comment (sxa) [inline get.sh:166]: These options to `find` are GNU specific so won't work on e.g. AIX/Solaris. I think this is currently set up to run on all platforms so I would expect it to fail on at least those two.

File: get.sh
Comment (andrew-m-leonard) [reply get.sh:452]: no I mean "brevity" as in: concise and exact use of words in writing or speech.
ie.easy to understand
one liners I find complex and hard to understand as they typically use options or commands i've never heard of or use, like "-r"
also a function keeps the logic the same in the several places it is used, rather than having to change in each line.

I understand this is subjective... but it's the way I prefer

Comment (sxa) [reply get.sh:452]: I would argue that `if [[ `ls -d */ | wc -l` -eq 1 ]] && [[ `find . -maxdepth 1 -mindepth 1 | wc -l` -eq 1 ]]; then` is particularly unclear and hard to follow exactly what it's doing, whereas `-r` to test if something is present is far more common, but as you say it's subjective ;-)

File: get.sh
Comment (sxa) [reply get.sh:164]: > should I call it "stripFolderIfArchiveContainsOnlyOne" ?

Up to you on the name - I don't mind too much.

File: get.sh
Comment (andrew-m-leonard) [reply get.sh:166]: good point :-)

Comment (andrew-m-leonard) [reply get.sh]: Changed to check current directory contains 1 directory and ONLY 1 directory..
Will run cross platform test and see if works, thanks

File: get.sh
Comment (andrew-m-leonard) [reply get.sh:164]: i'll leave name as is

**Comment by andrew-m-leonard** [comment]
Grinder on all platforms using "upstream" job output:
   -  "x64Mac" : https://ci.adoptium.net/job/Grinder/11440/
   -  "x64Linux" : 
   -  "x64AlpineLinux" : 
   -  "aarch64AlpineLinux" : 
   - "x64Windows" : 
   - "aarch64Windows" : 
   - "x32Windows" : 
   - "ppc64Aix" : 
   - "ppc64leLinux" : 
   - "s390xLinux" : 
   - "aarch64Linux" : 
   - "aarch64Mac" : 
   - "arm32Linux" : 
   - "riscv64Linux" : 
   - "x64Solaris" : 
   - "sparcv9Solaris" :

**Comment by andrew-m-leonard** [comment]
> * [windows](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44961/) - unzip passed
> 
>     * [vmfarm xlinux](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44962/) - unzip passed
> 
>     * [vmfarm windows](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44963/) - unzip passed
> 
>     * [openj9 windows](https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/3989/) - unzip passed
> 
>     * [zos fyrec21g](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder_testList_0/2346/) - unzip passed
> 
>     * [zos](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/44986/)

Thank you @llxia

File: get.sh
Comment (andrew-m-leonard) [reply get.sh]: all pass now https://github.com/adoptium/aqa-tests/pull/5750#issuecomment-2483268215

File: get.sh
Comment (sxa) [inline get.sh:178]: Assuming this is just trying to trap the case where fewer than two parameters are provided, this seems less complex
```suggestion
	if [ $# -lt 2 ]; then
```

File: get.sh
Comment (sxa) [inline get.sh:419]: Maybe be explicit as to why this is here
```suggestion
				# Remove 1 possibly 2 top-level folders (debugimage has 2)
```

File: get.sh
Comment (sxa) [inline get.sh:167]: This should ideally include `-1` to explicitly put each entry on a separate line as input to `wc`, although it seems to work regardless (surprisingly - I guess `ls` is operating differently when it's in a pipe)

```suggestion
        if [[ `ls -1d */ | wc -l` -eq 1 ]] && [[ `ls | wc -l` -eq 1 ]]; then
```

Comment (sxa) [inline get.sh:167]: Alternate option which seems much more readable IMHO
```suggestion
        if [[ $(ls -1 | wc -l) -eq 1 ]]; then
          folder=$(ls -d */)
            if [ -d "$folder" ]; then
```
Noting that I'm suggesting $(...) instead of backticks because the linter typically objects to those - I'm surprised it hasn't objected in this case as it's picked me up on it quite a lot (although I personally prefer the backticks!)

**Review by sxa — APPROVED** [review]
Approving, but with a number of suggested updates for clarity

File: get.sh
Comment (andrew-m-leonard) [reply get.sh:167]: I used them to be consistent with what's used in this method already, but yeah usually $()

Comment (andrew-m-leonard) [reply get.sh]: done

**Comment by andrew-m-leonard** [comment]
Sanity check with latest review updates: https://ci.adoptium.net/job/Grinder/11496/

**Comment by sxa** [comment]
> I think we're good now @ShelleyLambert ?

I think you want @smlambert :-)

**Comment by smlambert** [comment]
@llxia - some additional changes have been made to this PR, if you wish to rerun a couple of your Grinders to verify the last of these changes are good, please do so.  I will lift the block on this PR once I get the thumbs up from you, as it LGTM.

**Comment by llxia** [comment]
[windows](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/45050/)

[vmfarm xlinux](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/45051/)

[vmfarm windows](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/45052/)

[openj9 windows](https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/3994/)

[zos fyrec21g](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder_testList_0/2353/)

[zos](https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/45049/)

**Review by llxia — APPROVED** [review]

**Review by smlambert — APPROVED** [review]
thanks @andrew-m-leonard and all 👍

**Pull request merged**
